src/fontra/views/editor/edit-tools-shape.js

```diff
@@ -0,0 +1,202 @@
+      return;
+    }
+
+    let mouseRect;
+    let eventTemp;
+    for await (const event of eventStream) {
+      eventTemp = event;
+      const point = this.sceneController.selectedGlyphPoint(event);
+      if (point.x === undefined) {
+        // We can receive non-pointer events like keyboard events: ignore
+        continue;
+      }
+      mouseRect = rectangle.rectRound({
+        xMin: initialPoint.x,
+        yMin: initialPoint.y,
+        xMax: point.x,
+        yMax: point.y,
+      });
+      const drawPath = new Path2D();
+      this.drawShapePath2D(drawPath, mouseRect, event);
+      this.sceneModel.shapeToolShapePath = drawPath;
+      this.sceneModel.event = event;
+      this.canvasController.requestUpdate();
+    }
+
+    delete this.sceneModel.shapeToolShapePath;
+    delete this.sceneModel.event;
+    this.canvasController.requestUpdate();
+
+    // rectsize return when too small
+    if (!mouseRect) {
+      return;
+    }
+
+    const pathNew = this.drawShapeVarPackedPath(mouseRect, eventTemp);
+    // reversed contour direction
+    if (eventTemp.ctrlKey) {
+      this.reversePath(pathNew);
+    }
+    this.addShapePath(pathNew);
+  }
+
+  async addShapePath(pathNew) {
+    await this.sceneController.editGlyphAndRecordChanges(
+      (glyph) => {
+        const editLayerGlyphs = this.sceneController.getEditingLayerFromGlyphLayers(
+          glyph.layers
+        );
+
+        const firstLayerGlyph = Object.values(editLayerGlyphs)[0];
+        const selection = new Set();
+        const firstIndex = firstLayerGlyph.path.numPoints;
+        for (const index of range(pathNew.numPoints)) {
+          const point = pathNew.getPoint(index);
+          if (!point.type) {
+            selection.add(`point/${firstIndex + index}`);
+          }
+        }
+        this.sceneController.selection = selection;
+
+        for (const [layerName, layerGlyph] of Object.entries(editLayerGlyphs)) {
+          layerGlyph.path.appendPath(pathNew);
+        }
+        return "add shape";
+      },
+      undefined,
+      true
+    );
+  }
+
+  deactivate() {
+    this.canvasController.requestUpdate();
+  }
+
+  drawShapePath2D(path2d, mouseRect, event) {
+    const path = this.drawShapeVarPackedPath(mouseRect, event);
+
+    return path.drawToPath2d(path2d);
+  }
+
+  drawShapeVarPackedPath(mouseRect, event) {
+    let x = mouseRect.xMin;
+    let y = mouseRect.yMin;
+    let width = mouseRect.xMax - mouseRect.xMin;
+    let height = mouseRect.yMax - mouseRect.yMin;
+
+    // make square, not rectangle
+    if (event.shiftKey) {
+      const size = (Math.abs(width) + Math.abs(height)) / 2;
+      width = size * Math.sign(width);
+      height = size * Math.sign(height);
+    }
+
+    // positon at center
+    if (event.altKey) {
+      x = x - width;
+      y = y - height;
+      width *= 2;
+      height *= 2;
+    }
+
+    return VarPackedPath.fromUnpackedContours(
+      this.getUnpackedContour(x, y, width, height)
```

Comment:
```suggestion
      this.getUnpackedContours(x, y, width, height)
```
Should be plural: it returns an array of contours.

Reply:
Done.